embassy/src/channel/pubsub.rs

```diff
@@ -0,0 +1,651 @@
+/// an error to indicate that it has lagged.
+pub struct PubSubChannel<M: RawMutex, T: Clone, const CAP: usize, const SUBS: usize, const PUBS: usize> {
+    inner: Mutex<M, RefCell<PubSubState<T, CAP, SUBS, PUBS>>>,
+}
+
+impl<M: RawMutex, T: Clone, const CAP: usize, const SUBS: usize, const PUBS: usize>
+    PubSubChannel<M, T, CAP, SUBS, PUBS>
+{
+    /// Create a new channel
+    pub const fn new() -> Self {
+        Self {
+            inner: Mutex::const_new(M::INIT, RefCell::new(PubSubState::new())),
+        }
+    }
+
+    /// Create a new subscriber. It will only receive messages that are published after its creation.
+    ///
+    /// If there are no subscriber slots left, an error will be returned.
+    pub fn subscriber(&self) -> Result<Subscriber<'_, T>, Error> {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+
+            // Search for an empty subscriber spot
+            for (i, sub_spot) in s.subscriber_wakers.iter_mut().enumerate() {
+                if sub_spot.is_none() {
+                    // We've found a spot, so now fill it and create the subscriber
+                    *sub_spot = Some(WakerRegistration::new());
+                    return Ok(Subscriber {
+                        subscriber_index: i,
+                        next_message_id: s.next_message_id,
+                        channel: self,
+                    });
+                }
+            }
+
+            // No spot was found, we're full
+            Err(Error::MaximumSubscribersReached)
+        })
+    }
+
+    /// Create a new publisher
+    ///
+    /// If there are no publisher slots left, an error will be returned.
+    pub fn publisher(&self) -> Result<Publisher<'_, T>, Error> {
+        self.inner.lock(|inner| {
+            let mut s = inner.borrow_mut();
+
+            // Search for an empty publisher spot
+            for (i, pub_spot) in s.publisher_wakers.iter_mut().enumerate() {
+                if pub_spot.is_none() {
+                    // We've found a spot, so now fill it and create the subscriber
+                    *pub_spot = Some(WakerRegistration::new());
+                    return Ok(Publisher {
+                        publisher_index: i,
+                        channel: self,
+                    });
+                }
+            }
+
+            // No spot was found, we're full
+            Err(Error::MaximumPublishersReached)
+        })
+    }
+
+    /// Create a new publisher that can only send immediate messages.
+    /// This kind of publisher does not take up a publisher slot.
+    pub fn immediate_publisher(&self) -> ImmediatePublisher<'_, T> {
+        ImmediatePublisher { channel: self }
+    }
+}
+
+impl<M: RawMutex, T: Clone, const CAP: usize, const SUBS: usize, const PUBS: usize> PubSubBehavior<T>
+    for PubSubChannel<M, T, CAP, SUBS, PUBS>
+{
+    fn try_publish(&self, message: T) -> Result<(), T> {
+        self.inner.lock(|inner| {
```

Comment:
all the functions here acquire the mutex internally, and you call them from the future/stream impls. (for example, a subscriber will do `get_message` then `register_subscriber_waker`). This has 2 issues:

1. It's slightly slower because you're acquiring the mutex twice. 
2. It can lead to race conditions, because "in between" of the two calls, another thread can sneak in and change the state. For example, when receiving in a subscriber: 
    - Call `get_message`, there are no messages so it returns None.
    - Another thread sneaks in and pushes a message
    - Call `register_subscriber_waker`. The waker gets registered, and NOT woken because it was registered AFTER the message came.
    - The subscriber deadlocks, it yields even though there's a message in the queue and the waker never gets woken.

To fix these, you could move all these fns into `PubSubState` taking `&mut self`, and make the futs acquire the mutex themselves.